### ISSUE TITLE
Create a zod type from a TS enum

### ISSUE THREAD

**Issue by chrbala** [vriad/zod#103]
Right now if you want to create a type from a TS enum, you have do something like the following:

```ts
enum PlantLifecycle {
  Evergreen = 'EVERGREEN',
  Deciduous = 'DECIDUOUS',
  SemiDeciduous = 'SEMI_DECIDUOUS'
};

const lifecycle = z.union([
  z.literal(PlantLifecycle.Deciduous),
  z.literal(PlantLifecycle.Evergreen),
  z.literal(PlantLifecycle.SemiDeciduous),
]);
```

There should be an easier way of doing this, like:
```ts
enum PlantLifecycle {
  Evergreen = 'EVERGREEN',
  Deciduous = 'DECIDUOUS',
  SemiDeciduous = 'SEMI_DECIDUOUS'
};

z.enum(PlantLifecycle);
```

**Comment by johncantrell97** [comment]
I'm also running into this problem.  I'm trying both these methods to create a Zod enum with an existing typescript MemberRole enum:

```
const memberRoles = Object.keys(MemberRole).map(roleKey => MemberRole[roleKey])
const MemberRoleEnum = z.enum(memberRoles)


const memberRoleLiterals = Object.keys(MemberRole).map(roleKey => z.literal(MemberRole[roleKey]))
const MemberRoleUnion = z.union(memberRoleLiterals)
```

Neither of these attempts make Typescript happy.  It would really be nice to not have to manage the list of enum values in multiple places.

**Comment by maneetgoyal** [comment]
Similar issue:

I have an object literal

```ts
const lit = {
 Alpha: “a”,
 Beta: “b”
} as const;
```

Want to derive a zod enum from it but getting TS error with:

```ts
const zodEnum = zod.enum([Object.values(lit)]);
```

I suppressed the TS error by asserting [string, ...string[]] but then the inferred type wasn’t that helpful.

**Comment by vriad** [comment]
Just implemented a version of this that works for both enums and `const` objects. Still writing tests for it, it'll ship with zod@1.11.

**Comment by maneetgoyal** [comment]
Thanks @vriad. Only recently started working with Zod. Was having trouble with union types with Yup, so made the switch and so far been very happy with this tool 👍.

No other naming suggestions from my end. But just wanted to ask if the following case will also be handled.

Have an object literal where I store some numeric constants (like molecular weights, etc).

```
const MolecularWeights = {
  NaOH: 10,
  CaOH: 20,
  ...
} as const;
```

And converting to Zod via:

```
export const MolecularWeightsZod = zod.union([
  zod.literal(FFAMolecularWeightConstants.NaOH),
  zod.literal(FFAMolecularWeightConstants.CaOH),
  ...
]);
```

When I tried using `zod.enum`, TSC was throwing errors (because my values are `number` instead of `string`) so went with `zod.union`. You may have already solved this, but just wanted to verify once.

**Comment by vriad** [comment]
Yep, that should work fine.

**Comment by vriad** [comment]
@chrbala @maneetgoyal @johncantrell97 

hey folks, I implemented this in zod@1.11 as `z.nativeEnum`. documented [here](https://github.com/vriad/zod#native-enums). give it a shot!
